File: frontend/lib/models/boat.dart
enum BoatStatus {
  available,
  rented,
  maintenance;

  /// 安全解析 BoatStatus
  static BoatStatus fromString(String? value) {
    if (value == null) return BoatStatus.available;
    return values.firstWhere(
      (e) => e.name == value,
      orElse: () => BoatStatus.available,
    );
  }
}

class Boat {
  final int id;
  final String name;
  final String? type;
  final BoatStatus status;
  final double rentalPrice;
  final String? imageUrl;
  final String? description;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Boat({
    required this.id,
    required this.name,
    this.type,
    required this.status,
    required this.rentalPrice,
    this.imageUrl,
    this.description,
    required this.createdAt,
    this.updatedAt,
  });

  /// 安全解析 double 类型
  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  /// 安全解析 DateTime 类型
  static DateTime _parseDateTime(dynamic value, {bool required = false}) {
    if (value == null) return required ? DateTime.now() : DateTime(1970);
    if (value is String) return DateTime.tryParse(value) ?? DateTime.now();
    return DateTime.now();
  }

  factory Boat.fromJson(Map<String, dynamic> json) {
    return Boat(
      id: json['id'] as int? ?? 0,
      name: json['name'] as String? ?? '',
      type: json['type'] as String?,
      status: BoatStatus.fromString(json['status'] as String?),
      rentalPrice: _parseDouble(json['rental_price']),
      imageUrl: json['image_url'] as String?,
      description: json['description'] as String?,
      createdAt: _parseDateTime(json['created_at'], required: true),
      updatedAt: json['updated_at'] != null
          ? _parseDateTime(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'type': type,
      'status': status.name,
      'rental_price': rentalPrice,
      'image_url': imageUrl,
      'description': description,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}

File: backend/app/models/boat.py
from sqlalchemy import Column, Integer, DateTime, Text, DECIMAL, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from sqlalchemy.types import TypeDecorator, String
from app.database import Base
import enum


class BoatStatus(str, enum.Enum):
    AVAILABLE = "available"
    RENTED = "rented"
    MAINTENANCE = "maintenance"


class CaseInsensitiveEnum(TypeDecorator):
    """支持大小写不敏感的枚举类型"""
    impl = String
    cache_ok = True

    def __init__(self, enum_class, **kwargs):
        self.enum_class = enum_class
        super().__init__(**kwargs)

    def load_dialect_impl(self, dialect):
        return dialect.type_descriptor(String(50))

    def process_bind_param(self, value, dialect):
        if value is None:
            return None
        return value.value

    def process_result_value(self, value, dialect):
        if value is None:
            return None
        # 大小写不敏感处理
        value_lower = value.lower()
        for member in self.enum_class:
            if member.value.lower() == value_lower:
                return member
        # 如果找不到匹配的枚举值，返回默认值
        return self.enum_class.AVAILABLE

    def __repr__(self):
        return f"CaseInsensitiveEnum({self.enum_class.__name__})"


class Boat(Base):
    __tablename__ = "boats"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(50), nullable=False)
    type = Column(String(50))
    status = Column(CaseInsensitiveEnum(BoatStatus), default=BoatStatus.AVAILABLE)
    rental_price = Column(DECIMAL(10, 2), default=0.00)
    image_url = Column(String(255))
    description = Column(Text)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    rentals = relationship("BoatRental", back_populates="boat")


class BoatRental(Base):
    __tablename__ = "boats_rentals"

    id = Column(Integer, primary_key=True, index=True)
    boat_id = Column(Integer, ForeignKey("boats.id"))
    user_id = Column(Integer, ForeignKey("users.id"))
    rental_time = Column(DateTime(timezone=True), nullable=False)
    return_time = Column(DateTime(timezone=True))
    status = Column(String(20), default="active")

    boat = relationship("Boat", back_populates="rentals")
    user = relationship("User")
